cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1679,6 +1704,15 @@ class AsofJoinNode : public ExecNode {
   const Ordering& ordering() const override { return ordering_; }
 
   Status InputReceived(ExecNode* input, ExecBatch batch) override {
+    // InputReceived may be called after execution was finished. Pushing it to the
+    // InputState may cause the BackPressureController to pause the input, causing a
+    // deadlock
```

Comment:
```suggestion
    // InputReceived may be called after execution was finished. Pushing it to the
    // InputState is unnecessary since we're done (and anyway may cause the
    // BackPressureController to pause the input, causing a deadlock), so drop it.
```

Do we still deadlock *with* this short circuit but *without* ForceShutdown etc?

Reply:
Yes, the `forceShutdown` is still necessary. there's nothing stopping this order of events:

1. We receive enough data to finish the as of join.
2. Right before we finish processing and shut down the worker thread, lots of batches come in from input A. Input A pauses
3. We shut down the thread, and input A can't be unpaused

But your comment change suggestions  sound good to me